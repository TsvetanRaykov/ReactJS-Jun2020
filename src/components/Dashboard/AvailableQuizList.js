import React, { useEffect, useState } from 'react'
import quizService from '../../services/quizService'
import {
	List,
	ListItem,
	ListItemAvatar,
	Avatar,
	ListItemText,
	Typography,
	makeStyles,
} from '@material-ui/core'
import Loader from '../Loader'
import { withRouter } from 'react-router-dom'

const useStyles = makeStyles((theme) => ({
	inline: {
		display: 'inline',
	},
}))

const AvalableQuizList = (props) => {
	const classes = useStyles()

	const { history } = props
	const [quizzes, setQuizzes] = useState([])
	const [loading, setLoading] = useState(false)

	useEffect(() => {
		setLoading(true)
		quizService
			.getAvailable()
			.then((data) => {
				setQuizzes(data)
			})
			.finally(() => setLoading(false))
	}, [])

	const handleQuizClick = (id) => {
		history.push(`/quiz/progress/${btoa(id)}`)
	}

	const renderQuizList = () => {
		if (!quizzes || quizzes.length === 0) {
			return (
				<Typography variant='caption'>
					There are no quizzes available for you so far
				</Typography>
			)
		}
		return (
			<List>
				{quizzes.map((data) => (
					<ListItem
						key={data.id}
						alignItems='flex-start'
						dense
						button
						onClick={() => handleQuizClick(data.id)}
					>
						<ListItemAvatar>
							<Avatar alt='author' src={data.authorImg}></Avatar>
						</ListItemAvatar>
						<ListItemText
							color='textPrimary'
							primary={data.title}
							secondary={
								<Typography>
									<Typography
										component='span'
										variant='body2'
										className={classes.inline}
										color='textPrimary'
									>
										{`${data.duration / 60} min.`}
									</Typography>
									{` â€” ${data.description}`}
								</Typography>
							}
						></ListItemText>
					</ListItem>
				))}
			</List>
		)
	}

	return loading ? <Loader /> : renderQuizList()
}

export default withRouter(AvalableQuizList)
