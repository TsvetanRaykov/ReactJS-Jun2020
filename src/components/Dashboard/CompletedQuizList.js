import React, { useEffect, useState } from 'react'
import quizService from '../../services/quizService'
import { withRouter } from 'react-router-dom'
import Loader from '../Loader'
import {
	Typography,
	makeStyles,
	Accordion,
	AccordionSummary,
	AccordionDetails,
	Box,
	TableContainer,
	Table,
	TableHead,
	TableRow,
	TableCell,
	TableBody,
	Paper,
} from '@material-ui/core'
import { ExpandMore } from '@material-ui/icons'
import { padZero, timeOptions } from '../../utils'

const useStyles = makeStyles((theme) => ({
	root: {},
	table: {},
	heading: {
		fontSize: theme.typography.pxToRem(15),
		fontWeight: theme.typography.fontWeightMedium,
	},
	subheading: {},
}))

const CompletedQuizList = (props) => {
	const [loading, setLoading] = useState(false)
	const [quizzes, setQuizzes] = useState({})
	const [expanded, setExpanded] = useState(false)
	const classes = useStyles()

	useEffect(() => {
		setLoading(true)
		quizService
			.getCompleted()
			.then((data) => {
				setQuizzes(data)
			})
			.finally(() => setLoading(false))
	}, [])

	const handleChange = (panel) => (_, isExpanded) => {
		setExpanded(isExpanded ? panel : false)
	}
	const renderQuizList = () => {
		return (
			<div className={classes.root}>
				{Object.keys(quizzes).length > 0 ? (
					Object.keys(quizzes).map((id) => {
						const { title, description, completions } = quizzes[id]
						return (
							<Accordion
								key={id}
								expanded={expanded === id}
								onChange={handleChange(id)}
							>
								<AccordionSummary expandIcon={<ExpandMore />}>
									<Box display='flex' flexDirection='column'>
										<p className={classes.heading}>{title}</p>
										<p className={classes.subheading}>{description}</p>
									</Box>
								</AccordionSummary>
								<AccordionDetails>
									<TableContainer component={Paper}>
										<Table
											className={classes.table}
											size='small'
											aria-label='a dense table'
										>
											<TableHead>
												<TableRow>
													<TableCell>Date</TableCell>
													<TableCell align='right'>duration</TableCell>
													<TableCell align='right'>correct</TableCell>
													<TableCell align='right'>total</TableCell>
													<TableCell align='right'>score</TableCell>
												</TableRow>
											</TableHead>
											<TableBody>
												{completions.map((row, i) => {
													const minutes = padZero(
														parseInt(row.duration / 60, 10)
													)
													const seconds = padZero(
														parseInt(row.duration % 60, 10)
													)
													return (
														<TableRow key={i}>
															<TableCell component='th' scope='row'>
																{row.passedAt
																	.toDate()
																	.toLocaleDateString('en', timeOptions)}
															</TableCell>
															<TableCell align='right'>{`${minutes} : ${seconds}`}</TableCell>
															<TableCell align='right'>{row.correct}</TableCell>
															<TableCell align='right'>{row.total}</TableCell>
															<TableCell align='right'>{row.score}</TableCell>
														</TableRow>
													)
												})}
											</TableBody>
										</Table>
									</TableContainer>
								</AccordionDetails>
							</Accordion>
						)
					})
				) : (
					<Typography variant='caption'>
						You haven't completed quizzes yet
					</Typography>
				)}
			</div>
		)
	}

	return loading ? <Loader /> : renderQuizList()
}

export default withRouter(CompletedQuizList)
