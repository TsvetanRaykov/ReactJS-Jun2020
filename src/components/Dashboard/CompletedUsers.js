import React from 'react'
import {
	Chip,
	Avatar,
	withStyles,
	Tooltip,
	Accordion,
	AccordionSummary,
	Typography,
	AccordionDetails,
	makeStyles,
} from '@material-ui/core'
import quizService from '../../services/quizService'
import { timeOptions } from '../../utils'
import { ExpandMore } from '@material-ui/icons'

const useStyles = makeStyles((theme) => ({
	root: {
		width: '100%',
	},
	heading: {
		fontSize: theme.typography.pxToRem(15),
		fontWeight: theme.typography.fontWeightMedium,
		marginLeft: theme.spacing(1),
	},
	head: {
		fontWeight: theme.typography.fontWeightMedium,
		backgroundColor: 'rgba(192,192,192,0.1)',
	},
}))

const HtmlTooltip = withStyles((theme) => ({
	tooltip: {
		backgroundColor: '#f5f5f9',
		color: 'rgba(0, 0, 0, 0.87)',
		fontSize: theme.typography.pxToRem(12),
		border: '1px solid #dadde9',
	},
}))(Tooltip)

const CompleteUser = (props) => {
	const classes = useStyles()
	const {
		setModalDialog,
		setLoading,
		loadOwnQuizes,
		completedBy,
		duration,
		id,
	} = props
	const handleReleaseComplete = (qid, uid, arr) => {
		setModalDialog({
			title: 'Are you sure?',
			message:
				'This will allow that person to try the quiz again. But will remove it from that list.',
			handleYes: () => {
				setModalDialog({ open: false })
				setLoading(true)
				quizService.releaseQuiz(
					qid,
					arr.filter((u) => u.uid !== uid)
				)
				loadOwnQuizes()
			},
			handleNo: () => {
				setModalDialog({ open: false })
			},
			open: true,
		})
	}

	return (
		<>
			{completedBy && completedBy.length > 0 && (
				<Accordion elevation={0} className={classes.root}>
					<AccordionSummary
						expandIcon={<ExpandMore />}
						className={classes.head}
					>
						<Typography className={classes.heading}>Completed by</Typography>
					</AccordionSummary>
					<AccordionDetails>
						{completedBy.map((p) => {
							const { correct, passedAt, timeLeft, total } = p.result

							const min = parseInt((duration - timeLeft) / 60, 10)
							const sec = parseInt((duration - timeLeft) % 60, 10)

							return (
								<HtmlTooltip
									key={p.uid}
									title={
										<>
											<em>
												{passedAt
													.toDate()
													.toLocaleDateString('en', timeOptions)}
											</em>
											<br />
											<em>Result:</em>&nbsp;
											<b>{`${correct}/${total}`}</b>
											<br />
											<em>Time:</em>&nbsp;
											<b>{`${min}:${sec}`}</b>
										</>
									}
								>
									<Chip
										avatar={<Avatar alt={p.email} src={p.photo} />}
										label={p.email}
										onDelete={() =>
											handleReleaseComplete(id, p.uid, completedBy.slice(0))
										}
										variant='outlined'
										color='primary'
									/>
								</HtmlTooltip>
							)
						})}
					</AccordionDetails>
				</Accordion>
			)}
		</>
	)
}

export default CompleteUser
