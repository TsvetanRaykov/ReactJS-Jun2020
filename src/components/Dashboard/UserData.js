import React, { useContext, createRef, useState, useEffect } from 'react'
import {
	Grid,
	List,
	ListItem,
	ListItemAvatar,
	Divider,
	ListItemText,
	CircularProgress,
	Tooltip,
	Avatar,
	withStyles,
	Link,
} from '@material-ui/core'
import { PersonRounded, EmailRounded } from '@material-ui/icons'
import EditableTextField from '../EditableTextField'
import UserContext from '../../Context'
import userService from '../../services/userService'

const styles = (theme) => ({
	root: {
		width: '100%',
		backgroundColor: theme.palette.background.paper,
	},
	avatar: {
		width: theme.spacing(15),
		height: theme.spacing(15),
		cursor: 'pointer',
	},
	file: {
		opacity: '0%',
		position: 'absolute',
	},
})

const UserData = (props) => {
	const { classes } = props
	const { updateUser } = useContext(UserContext)
	const [userName, setUserName] = useState('')
	const [userImg, setUserImg] = useState('')
	const [userEmail, setUserEmail] = useState('')

	useEffect(() => {
		setUserImg(userService.getCurrentUser().userImg)
		setUserName(userService.getCurrentUser().userName)
		setUserEmail(userService.getCurrentUser().userEmail)
	}, [])

	const fileRef = createRef()

	const imageHandler = (event) => {
		if (event.target.files.length === 0) {
			return
		}
		const imageFile = event.target.files[0]
		const reader = new FileReader()

		reader.onload = () => {
			if (reader.readyState === 2) {
				setUserImg(reader.result)
				updateUser({
					userImg: reader.result,
				})
				userService.addImage(imageFile)
			}
		}

		reader.readAsDataURL(imageFile)
	}

	const triggerInputFile = () => {
		if (fileRef.current !== undefined && fileRef.current.click !== undefined) {
			fileRef.current.click()
		}
	}

	const saveName = (newName) => {
		updateUser({ userName: newName })
		userService.updateName(newName)
	}
	return (
		<Grid container>
			<Grid item xs={4} md={3} lg={2}>
				{userImg ? (
					<Tooltip
						title={
							<>
								Is that you? <br />
								Click to change the picture.
							</>
						}
					>
						<Avatar
							className={classes.avatar}
							alt={userName}
							src={userImg}
							onClick={triggerInputFile}
						/>
					</Tooltip>
				) : (
					<CircularProgress />
				)}
				<input
					id='img-input'
					accept='image/*'
					type='file'
					className={classes.file}
					onChange={imageHandler}
					ref={fileRef}
				/>
			</Grid>
			<Grid item xs={8} md={9} lg={10}>
				<List className={classes.root}>
					<ListItem>
						<ListItemAvatar>
							<PersonRounded color='primary' />
						</ListItemAvatar>
						<EditableTextField text={userName} callback={saveName} />
					</ListItem>
					<Divider />
					<Link href={`mailto://${userEmail}`}>
						<ListItem>
							<ListItemAvatar>
								<EmailRounded color='primary' />
							</ListItemAvatar>
							<ListItemText primary={userEmail} />
						</ListItem>
					</Link>
				</List>
			</Grid>
		</Grid>
	)
}

export default withStyles(styles)(UserData)
