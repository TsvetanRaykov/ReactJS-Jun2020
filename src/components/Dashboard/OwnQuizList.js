import React, { useState, useEffect, useContext } from 'react'
import { withRouter } from 'react-router-dom'
import quizService from '../../services/quizService'
import { padZero } from '../../utils'
import Loader from '../Loader'
import {
	withStyles,
	Accordion,
	AccordionSummary,
	Typography,
	AccordionDetails,
	Chip,
	Box,
	AccordionActions,
	Button,
} from '@material-ui/core'
import {
	ExpandMore,
	Public,
	VpnLock,
	DeleteForever,
	Settings,
	AccessTime,
} from '@material-ui/icons'
import ModalDialog from '../../components/shared/ModalDialog'
import Context from '../../Context'
import CompletedUsers from './CompletedUsers'

const styles = (theme) => ({
	root: {
		width: '100%',
	},
	heading: {
		fontSize: theme.typography.pxToRem(15),
		fontWeight: theme.typography.fontWeightMedium,
		marginLeft: theme.spacing(1),
	},
	label: {
		'& span': {
			color: theme.palette.text.primary,
		},
	},

	button: {
		minWidth: '100px',
	},
	duration: {
		display: 'inline-block',
		fontWeight: 'bold',
		margin: theme.spacing(0, 1, 0, 0),
		minWidth: '50px',
	},
})

const OwnQuizList = (props) => {
	const { classes } = props
	const [quizzes, setQuizzes] = useState([])
	const [loading, setLoading] = useState(true)
	const [expanded, setExpanded] = useState(false)
	const { updateQuiz, updateQuizSnapshot } = useContext(Context)

	const [modalDialog, setModalDialog] = useState({
		title: '',
		message: '',
		open: false,
		handleYes: () => {},
		handleNo: () => {},
	})

	const loadOwnQuizes = () => {
		quizService
			.getPersonal()
			.then((data) => {
				setQuizzes(data)
			})
			.finally(() => setLoading(false))
	}

	useEffect(() => {
		loadOwnQuizes()
	}, [])

	const renderChip = (isPublic) => {
		return isPublic ? (
			<Chip
				className={classes.label}
				variant='outlined'
				color='primary'
				size='small'
				icon={<Public />}
				label='public'
			/>
		) : (
			<Chip
				className={classes.label}
				variant='outlined'
				color='secondary'
				size='small'
				icon={<VpnLock />}
				label='private'
			/>
		)
	}

	const handleEditClick = (id) => {
		quizService.getById(id).then(({ data }) => {
			updateQuiz({ ...data, id })
			updateQuizSnapshot({ ...data, id })
			props.history.push('/quiz/edit/questions')
		})
	}

	const handleDeleteClick = (id) => {
		setModalDialog({
			title: 'Are you sure?',
			handleYes: () => {
				setModalDialog({ open: false })
				setLoading(true)
				quizService.deleteQuiz(id).finally(() => loadOwnQuizes())
			},
			handleNo: () => {
				setModalDialog({ open: false })
			},
			open: true,
		})
	}

	const handleChange = (panel) => (_, isExpanded) => {
		setExpanded(isExpanded ? panel : false)
	}

	const renderQuizzes = () => {
		return (
			<div className={classes.root}>
				{quizzes.length > 0 ? (
					quizzes.map((data) => {
						const { title, description, isPublic, duration, completedBy, id } =
							data
						const min = padZero(duration / 60)
						const sec = padZero(duration % 60)

						return (
							<Accordion
								key={id}
								expanded={expanded === id}
								onChange={handleChange(id)}
							>
								<AccordionSummary expandIcon={<ExpandMore />}>
									{renderChip(isPublic)}
									<Typography className={classes.heading}>{title}</Typography>
								</AccordionSummary>
								<AccordionDetails>
									<Box display='flex' alignItems='center'>
										<AccessTime />
										&nbsp;
										<div className={classes.duration}>{`${min} : ${sec} `}</div>
										{description}
									</Box>
								</AccordionDetails>
								<AccordionActions>
									<Button
										data-cy={`delete-${title}`}
										size='small'
										variant='contained'
										color='secondary'
										className={classes.button}
										startIcon={<DeleteForever />}
										onClick={() => handleDeleteClick(id)}
									>
										Delete
									</Button>
									<Button
										size='small'
										variant='contained'
										color='primary'
										className={classes.button}
										startIcon={<Settings />}
										onClick={() => handleEditClick(id)}
									>
										Edit
									</Button>
								</AccordionActions>
								<Box display='flex' alignItems='center'>
									<CompletedUsers
										completedBy={completedBy}
										setModalDialog={setModalDialog}
										loadOwnQuizes={loadOwnQuizes}
										setLoading={setLoading}
										duration={duration}
										id={id}
									/>
								</Box>
							</Accordion>
						)
					})
				) : (
					<Typography variant='caption'>
						You haven't created quizzes yet
					</Typography>
				)}
				<ModalDialog {...modalDialog} />
			</div>
		)
	}

	return loading ? <Loader /> : renderQuizzes()
}

export default withRouter(withStyles(styles)(OwnQuizList))
