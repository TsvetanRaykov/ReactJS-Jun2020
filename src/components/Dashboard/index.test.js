import { shallow } from 'enzyme'
import React from 'react'
// import { createRender } from 'react-test-renderer'
import { act, create } from 'react-test-renderer'
import TestWrapper from '../../utils/test-wrapper'
import AvailableQuizList from './AvailableQuizList'
import CompletedQuizList from './CompletedQuizList'
import CompletedUsers from './CompletedUsers'
import OwnQuizList from './OwnQuizList'
import UserData from './UserData'

jest.mock('react', () => ({
	...jest.requireActual('react'),
	useLayoutEffect: jest.requireActual('react').useEffect,
}))

describe('<AvailableQuizList />', () => {
	it('Should match the snapshot', () => {
		act(() => {
			const wrapper = shallow(
				<TestWrapper>
					<AvailableQuizList />)
				</TestWrapper>
			)
			expect(wrapper).toMatchSnapshot()
		})
	})
})

describe('<CompletedQuizList />', () => {
	it('Should match the snapshot', () => {
		act(() => {
			const wrapper = shallow(
				<TestWrapper>
					<CompletedQuizList />)
				</TestWrapper>
			)
			expect(wrapper).toMatchSnapshot()
		})
	})
})

describe('<CompletedUsers />', () => {
	it('Should match the snapshot', () => {
		act(() => {
			const wrapper = create(
				<TestWrapper>
					<CompletedUsers />)
				</TestWrapper>
			).toJSON()
			expect(wrapper).toMatchSnapshot()
		})
	})
})

describe('<OwnQuizList />', () => {
	it('Should match the snapshot', () => {
		act(() => {
			const wrapper = shallow(
				<TestWrapper>
					<OwnQuizList />)
				</TestWrapper>
			)
			expect(wrapper).toMatchSnapshot()
		})
	})
})

describe('<UserData />', () => {
	it('Should match the snapshot', () => {
		act(() => {
			const wrapper = shallow(
				<TestWrapper>
					<UserData />)
				</TestWrapper>
			)
			expect(wrapper).toMatchSnapshot()
		})
	})
})
