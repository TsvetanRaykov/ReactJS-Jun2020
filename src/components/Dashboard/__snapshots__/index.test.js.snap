// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<AvailableQuizList /> Should match the snapshot 1`] = `ShallowWrapper {}`;

exports[`<CompletedQuizList /> Should match the snapshot 1`] = `ShallowWrapper {}`;

exports[`<CompletedUsers /> Should match the snapshot 1`] = `null`;

exports[`<OwnQuizList /> Should match the snapshot 1`] = `ShallowWrapper {}`;

exports[`<UserData /> Should match the snapshot 1`] = `ShallowWrapper {}`;
